### PR TITLE
Backport `RadixTree.h` and improve PDB generation

### PR DESCRIPTION
**Pull request by aleclearmind**
I'm making some changes to the LLVM repo.

Two set of changes:
1. Add `RadixTree.h` from upstream, I need it on my branch and might useful in the future. Most of these are cheery-picks (plus some merge conflicts resolved):
    ```
    [RadixTree] Use std::optional for Node::Value (#165299)
    [ADT] Remove #include <limits> in RadixTree.h (NFC) (#165115)
    [ADT][NFC] Add missing #include <vector> (#165068)
    [NFC][ADT] Add RadixTree (#164524)
    [ADT] Add llvm::mismatch and llvm::range_size range wrappers
    [ADT] Allow reverse to find free rbegin/rend functions (#87840)
    [NFC][ADT] Add reverse iterators and `value_type` to StringRef (#105579)
    [ADT] Add deduction guide for iterator_range
    ```

   These changes are close to the tip of the branch

2. Make changes to `clang-release` to provide richer debugging information in PDB (used by `win32metadata`).

    ```
    Add LF_ALIAS support to CodeView type system
    Retain declaration-only subprograms and emit LF_FUNC_ID in CodeView
    Add CLANG_DEBUG_INFO_ALLOWED_FILES filter for debug info emission
    Add -fcase-insensitive-paths flag
    ```

   These are changes closer to the beginning of the branch: they will be included in the `clang-release` component (and the corresponding tag).

   These are all custom commits except the `LF_ALIAS` one which is adapted from [PR #153936](https://github.com/llvm/llvm-project/pull/153936).

   We could have a separate clang build for this set of changes, but they seem inncuous enough that it looks overkill to me.

Tests pass.

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -1402,6 +1402,10 @@ defm borland_extensions : BoolFOption<"borland-extensions",
   LangOpts<"Borland">, DefaultFalse,
   PosFlag<SetTrue, [CC1Option], "Accept non-standard constructs supported by the Borland compiler">,
   NegFlag<SetFalse>>;
+def fcase_insensitive_paths : Flag<["-"], "fcase-insensitive-paths">,
+  Group<f_Group>, Flags<[CC1Option]>,
+  HelpText<"Retry failed file lookups with a case-insensitive directory scan">,
+  MarshallingInfoFlag<FileSystemOpts<"CaseInsensitivePaths">>;
```
Comment (pfez) [inline clang/include/clang/Driver/Options.td]: I'd put this close to `def working_directory` below, which is the only other file system option.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -6551,6 +6551,9 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   if (IsMSVCCompat)
     CmdArgs.push_back("-fms-compatibility");
 
+  if (Args.hasArg(options::OPT_fcase_insensitive_paths))
+    CmdArgs.push_back("-fcase-insensitive-paths");
+
```
Comment (pfez) [inline clang/lib/Driver/ToolChains/Clang.cpp]: Any specific reason why this snippet must stay here? Otherwise I'd put after line 5745, where the the code handles `OPT_working_directory`, which is the only other file system option.

File: clang/lib/Basic/FileManager.cpp [hunk 3]
```diff
@@ -587,15 +621,30 @@ FileManager::getStatValue(StringRef Path, llvm::vfs::Status &Status,
...
-                                  StatCache.get(), *FS);
+  auto EC = FileSystemStatCache::get(FilePath.c_str(), Status, isFile, F,
+                                     StatCache.get(), *FS);
+  if (!EC)
+    return EC;
+  if (FileSystemOpts.CaseInsensitivePaths)
+    if (auto Resolved = findPathCaseInsensitive(FilePath))
+      return FileSystemStatCache::get(*Resolved, Status, isFile, F,
+                                      StatCache.get(), *FS);
+  return EC;
```
Comment (pfez) [inline clang/lib/Basic/FileManager.cpp]: Duplicated code. The only changing thing is `Path`/`FilePath`. Make a lambda or a private method.

File: clang/lib/Basic/FileManager.cpp [hunk 4]
```diff
@@ -577,6 +577,40 @@ FileManager::getBufferForFileImpl(StringRef Filename, int64_t FileSize,
...
+/// Walk a path component-by-component, resolving each level with a
+/// case-insensitive directory scan.  Returns the real-cased path on success.
+static std::optional<std::string>
+findPathCaseInsensitive(StringRef CaseInsensitivePath) {
+  using llvm::sys::fs::directory_iterator;
+
+  SmallVector<StringRef> Components;
+  CaseInsensitivePath.split(Components, "/", -1, /*KeepEmpty=*/false);
+  if (Components.empty())
+    return std::nullopt;
```
Comment (pfez) [inline clang/lib/Basic/FileManager.cpp]: Instead of splitting the path like this I'd go with `llvm::sys::path::begin` and `llvm::sys::path::end`.

File: clang/lib/Basic/FileManager.cpp [hunk 5]
```diff
@@ -577,6 +577,40 @@ FileManager::getBufferForFileImpl(StringRef Filename, int64_t FileSize,
...
+
+  // Start from "/" for absolute paths, "" (cwd) for relative ones.
+  SmallString<256> Resolved;
+  if (CaseInsensitivePath.starts_with("/"))
```
Comment (pfez) [inline clang/lib/Basic/FileManager.cpp]: And `llvm::sys::path::is_absolute` for this.

File: clang/lib/Basic/FileManager.cpp [hunk 6]
```diff
@@ -577,6 +577,40 @@ FileManager::getBufferForFileImpl(StringRef Filename, int64_t FileSize,
...
+    Resolved = "/";
```
Comment (pfez) [inline clang/lib/Basic/FileManager.cpp]: And `Resolved = llvm::sys::path::root_path(CaseInsensitivePath)` here.

File: clang/lib/Basic/FileManager.cpp [hunk 7]
```diff
@@ -577,6 +577,40 @@ FileManager::getBufferForFileImpl(StringRef Filename, int64_t FileSize,
...
+
+  for (StringRef Component : Components) {
+    std::error_code EC;
+    bool Found = false;
+    for (directory_iterator It(Resolved, EC), End; It != End && !EC;
+         It.increment(EC)) {
+      StringRef EntryName = llvm::sys::path::filename(It->path());
+      if (EntryName.equals_insensitive(Component)) {
```
Comment (pfez) [inline clang/lib/Basic/FileManager.cpp:600]: Ah scary, we can have more than one hit here. But given the intended use case for this I don't see an obvious workaround. Can we print a warning here?

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 8]
```diff
@@ -6276,6 +6279,47 @@ void CodeGenModule::EmitDeclContext(const DeclContext *DC) {
...
+struct AllowedFilesFilter {
+  bool Loaded = false;
+  bool HasFilter = false;
+  llvm::StringSet<> Files;
+
+  void ensureLoaded() {
+    if (Loaded)
+      return;
+    Loaded = true;
+    if (auto Value = llvm::sys::Process::GetEnv("CLANG_DEBUG_INFO_ALLOWED_FILES")) {
```
Comment (pfez) [inline clang/lib/CodeGen/CodeGenModule.cpp]: Why an environment variable for this as and not a CLI option as well?

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 9]
```diff
@@ -6276,6 +6279,47 @@ void CodeGenModule::EmitDeclContext(const DeclContext *DC) {
...
+
+  bool hasFilter() { ensureLoaded(); return HasFilter; }
+
+  bool contains(const Decl *D, ASTContext &Context) {
+    ensureLoaded();
+    if (!HasFilter)
+      return true;
+    auto Loc = Context.getSourceManager().getExpansionLoc(D->getLocation());
+    auto Filename = Context.getSourceManager().getFilename(Loc);
+    return !Filename.empty() && Files.count(Filename.lower());
```
Comment (pfez) [inline clang/lib/CodeGen/CodeGenModule.cpp]: `Filename.lower()`? Are you assuming `OPT_fcase_insensitive_paths` here?

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 10]
```diff
@@ -6295,6 +6339,23 @@ void CodeGenModule::EmitTopLevelDecl(Decl *D) {
...
+    if (TheAllowedFilesFilter.hasFilter()) {
+      if (CGDebugInfo *DI = getModuleDebugInfo()) {
+        auto *FD = cast<FunctionDecl>(D);
+        if (!isa<CXXMethodDecl>(FD) &&
+            !FD->doesThisDeclarationHaveABody() &&
+            !FD->isDependentContext() && !FD->isImplicit() &&
+            isDeclInAllowedFiles(D, getContext()))
+          if (auto *FPT = FD->getType()->getAs<FunctionProtoType>())
+            DI->EmitFunctionDecl(GlobalDecl(FD), FD->getLocation(),
+                                 QualType(FPT, 0));
```
Comment (pfez) [inline clang/lib/CodeGen/CodeGenModule.cpp]: 🤔 this is confusing.

`CLANG_DEBUG_INFO_ALLOWED_FILES` is advertised at line 6282 as restricting the set of functions whose debug info are emitted. So I would expect that this environment variable only has the effect of reducing the number of debug info that are generated.

But here where when the environment variable is set you are effectively emitting more debug info than if it wasn't set (unlike all the other cases below, where you properly emit less debug info.).

At the very least the comment at line 6282. From my understanding of what you're trying to do with debug symbols, and from what I extrapolate by the code and comment here, the environment variable is actually meant to *force* (not to *allow*) the emission of debug symbols of functions that come from the list of allowed files, but only them and not others.

Can you elaborate?

Comment (pfez) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Also, all of this would be much clearer if we had a couple of tests for expected behavior and to exclude undesirable behavior.

**Review by pfez — COMMENTED** [review]
1. The first set of changes about radix tree is very much non controversial.
2. I flushed my comments to your changes.
3. I'm still looking in the commit you backported from upstream, given that it hasn't been merged yet.

**Comment by pfez** [comment]
The backport seems legit as well.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 10 above)
Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp]: This looks terrible indeed.

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (aleclearmind) [reply clang/include/clang/Driver/Options.td]: Done.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (aleclearmind) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Done.

File: clang/lib/Basic/FileManager.cpp (hunk 3 above)
Comment (aleclearmind) [reply clang/lib/Basic/FileManager.cpp]: Factored into a lambda.

File: clang/lib/Basic/FileManager.cpp (hunk 4 above)
Comment (aleclearmind) [reply clang/lib/Basic/FileManager.cpp]: Done.

File: clang/lib/Basic/FileManager.cpp (hunk 5 above)
Comment (aleclearmind) [reply clang/lib/Basic/FileManager.cpp]: Reworked the code.

File: clang/lib/Basic/FileManager.cpp (hunk 6 above)
Comment (aleclearmind) [reply clang/lib/Basic/FileManager.cpp]: Reworked the code

File: clang/lib/Basic/FileManager.cpp (hunk 7 above)
Comment (aleclearmind) [reply clang/lib/Basic/FileManager.cpp:600]: Added a warning

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 8 above)
Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Dropped the environment variable in favor of a new flag.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 9 above)
Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Fixed

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 10 above)
Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Removed the ad-hoc code, added a test.

Comment (pfez) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Better, and the tests definitely help understanding what's the intended outcome.

Speaking of the tests: I still find confusing that passing `-fdebug-info-allowed-file=allowed.h` does *at the same time* restrict the set of types for which you emit debug symbols *and* enlarge the set of functions for which you emit debug symbols.

I think that ideally the best thing to do would be to have 2 separate flags:
* `-ftype-debug-info-allowed-file` that only restrict the set of types for which you emit debug symbols to the allowed files, without adding additional debug symbols for functions that wouldn't otherwise have them
* `-fforce-function-debug-info-file` that only forces the emission of debug symbols for functions in the forced files, without adding the debug symbols to types that wouldn't otherwise have them

This is slightly more complicated than what you did, but it's less confusing, both from a user perspective and when you read the code. I still find it odd when reading the code that a flag called *allowed*-something ends up generating *more*. And I think I would find it odd as a user as well.

Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp]: The intent was to emit it unconditionally (even in absence of `-fdebug-info-allowed-file`).
I'd do something like this:

```c++
if (!isDeclInAllowedFiles(FD, CGM))
  return;
```

And add a note as the side-effect this commit has.

Comment (pfez) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Let me get this straight. You want to change the baseline functionality so that the additional debug info for the functions are always emitted even in absence of `-fdebug-info-allowed-file`?

If so, than make this in a separate commit, before the commit that adds `-fdebug-info-allowed-file`.

Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp]: > Let me get this straight. You want to change the baseline functionality so that the additional debug info for the functions are always emitted even in absence of -fdebug-info-allowed-file?

Yes.

> If so, than make this in a separate commit, before the commit that adds -fdebug-info-allowed-file.

OK.

Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp]: OK, we now have a commit with a test just for the emission of the extra debug info:

```
Retain decl-only subprograms in CodeView
```

Decoupled from the rest.

Comment (pfez) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Nice! Thanks

**Comment by aleclearmind** [comment]
@pfez I've made some changes due to some test failures in building clang-release:

* `Silence noisy profiling counter warning`: suppresses some profiling error messages.
  When building the profiling data for the PGO build of `clang-release`, sometimes we get these messages which make the test fail. We used to ignore these errors, since the reason we were running those tests is just to collect profiling data, more than actually testing things.
  Anyway, this fixes that and we should be able to drop `|| true` from [here](https://github.com/revng/orchestra/blob/develop/.orchestra/config/components/clang-release.yml#L99).
* `Retain decl-only subprograms in CodeView`
  1. We now exclude functions with a body, i.e., definitions (their debug information is built elsewhere).
  2. We now exclude functions for which the full type is not available (not deduced yet). This is related to templates and we don't care about them anyway.
  3. In `CGDebugInfo::EmitFunctionDecl`, we now invoke `createFunction` with `nullptr` in the new code path we introduced. The original code was designed to emit function debug info due to call sites, referencing a function debug info defined elsewhere. Here we're defining the function debug info itself, so there's nothing to reference. If we do, we get a recursive function debug info which crashes LLVM when debug info are being emitted.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 11]
```diff
@@ -6276,6 +6276,24 @@ void CodeGenModule::EmitDeclContext(const DeclContext *DC) {
...
+static void emitFunctionDecl(CodeGenModule &CGM, const FunctionDecl *FD) {
+  // Only emit for declaration-only functions; definitions already get
+  // a DISubprogram through the normal codegen path.
+  if (FD->doesThisDeclarationHaveABody())
+    return;
+  // Skip declarations whose type can't be represented in debug info
+  // (e.g. undeduced auto return types, dependent template types).
+  const auto *FPT = FD->getType()->getAs<FunctionProtoType>();
+  if (!FPT || FPT->isDependentType() || FPT->isUndeducedType())
+    return;
```
Comment (pfez) [inline clang/lib/CodeGen/CodeGenModule.cpp:6291]: Please test these two corner cases as well in the unit test.

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 12]
```diff
@@ -4219,10 +4220,17 @@ void CGDebugInfo::EmitFunctionDecl(GlobalDecl GD, SourceLocation Loc,
 
   llvm::DINodeArray Annotations = CollectBTFDeclTagAnnotations(D);
   llvm::DISubroutineType *STy = getOrCreateFunctionType(D, FnType, Unit);
+  // Only link to a prior declaration when creating a call-site entry.
+  // For standalone declarations the SP *is* the declaration; passing
+  // getFunctionDeclaration() here would let DISubprogram::get() unique
+  // the new node to the existing member SP, producing a self-referencing
+  // declaration link that causes infinite recursion in the DWARF emitter.
+  llvm::DISubprogram *Decl =
+      IsDeclForCallSite ? getFunctionDeclaration(D) : nullptr;
```
Comment (pfez) [inline clang/lib/CodeGen/CGDebugInfo.cpp:4229]: This bug fix should be in its own separate commit.

Comment (pfez) [reply clang/lib/CodeGen/CGDebugInfo.cpp:4229]: :thinking: on the other hand. Before the code you introduced this was effectively impossible to trigger. So leave it like this.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 11 above)
Comment (aleclearmind) [reply clang/lib/CodeGen/CodeGenModule.cpp:6291]: I can only test it doesn't crash. There's no "expected output" or "output that should not be there". Either it's like this code is not there, or we get a crash.

Comment (pfez) [reply clang/lib/CodeGen/CodeGenModule.cpp:6291]: Fair enough. Thanks for clarifying.

**Comment by pfez** [comment]
Merged in 722c9691a480134b35c7259bab1909a22cf294ca
